src/Umbraco.Core/Persistence/Repositories/EntityRepository.cs

```diff
@@ -49,6 +50,124 @@ internal Guid UnitKey
 
         #region Query Methods
 
+        public IEnumerable<IUmbracoEntity> GetPagedResultsByQuery(IQuery<IUmbracoEntity> query, Guid objectTypeId, long pageIndex, int pageSize, out long totalRecords,
+            string orderBy, Direction orderDirection, IQuery<IUmbracoEntity> filter = null)
+        {   
+            bool isContent = objectTypeId == new Guid(Constants.ObjectTypes.Document);
+            bool isMedia = objectTypeId == new Guid(Constants.ObjectTypes.Media);
+            var factory = new UmbracoEntityFactory();
+            
+            var sqlClause = GetBaseWhere(GetBase, isContent, isMedia, sql =>
+            {
+                if (filter != null)
+                {
+                    foreach (var filterClaus in filter.GetWhereClauses())
+                    {
+                        sql.Where(filterClaus.Item1, filterClaus.Item2);
+                    }
+                }
+            }, objectTypeId);
+            var translator = new SqlTranslator<IUmbracoEntity>(sqlClause, query);
+            var entitySql = translator.Translate();
+            var pagedSql = entitySql.Append(GetGroupBy(isContent, isMedia, false)).OrderBy("umbracoNode.id");
+
+            IEnumerable<IUmbracoEntity> result;
+
+            if (isMedia)
+            {
+                //Treat media differently for now, as an Entity it will be returned with ALL of it's properties in the AdditionalData bag!
+                var pagedResult = _work.Database.Page<dynamic>(pageIndex + 1, pageSize, pagedSql);
+                
+                var ids = pagedResult.Items.Select(x => (int) x.id).InGroupsOf(2000);
+                var entities = pagedResult.Items.Select(factory.BuildEntityFromDynamic).Cast<IUmbracoEntity>().ToList();
+
+                //Now we need to merge in the property data since we need paging and we can't do this the way that the big media query was working before
+                foreach (var idGroup in ids)
+                {
+                    var propSql = GetPropertySql(Constants.ObjectTypes.Media)
+                        .Where("contentNodeId IN (@ids)", new {ids = idGroup})
+                        .OrderBy("contentNodeId");
+
+                    //This does NOT fetch all data into memory in a list, this will read
+                    // over the records as a data reader, this is much better for performance and memory,
+                    // but it means that during the reading of this data set, nothing else can be read
+                    // from SQL server otherwise we'll get an exception.

```

Comment:
This will not cause any issues. The difference between Fetch<T> and Query<T> is that:

* Fetch - will open a DB reader and read from start to the end and hydrating objects for each row and allocating them all into memory in a List<T>
* Query - will open a DB reader and the foreach loop will read forward over the reader, it will only allocate one row and hydrate one item into memory at once - The only issue with this is if you try to perform another query inside of the foreach loop, this would throw an exception because you already have a data reader open

Query will allocate a lot less memory and should always be used when it makes sense instead of allocating all rows to mem. And yes, we do this elsewhere :)